MCprep_addon/spawner/models.py

```diff
@@ -0,0 +1,1350 @@
+# ##### BEGIN GPL LICENSE BLOCK #####
+#
+#  This program is free software; you can redistribute it and/or
+#  modify it under the terms of the GNU General Public License
+#  as published by the Free Software Foundation; either version 2
+#  of the License, or (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, write to the Free Software Foundation,
+#  Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# ##### END GPL LICENSE BLOCK #####
+
+
+#import math
+import os
+#import random
+#import time
+
+import bpy
+#import mathutils
+
+# addon imports
+from .. import conf
+from ..materials import generate
+from .. import util
+from .. import tracking
+
+
+# -----------------------------------------------------------------------------
+# Mesh swap functions
+# -----------------------------------------------------------------------------
+
+model_cache = {}
+model_cache_path = None
+
+def get_model_cache(context, clear=False):
+	"""Load groups/objects from model spawning lib if not cached, return key vars."""
+	global model_cache
+	global model_cache_path # used to auto-clear path if bpy prop changed
+
+	model_path = context.scene.model_path
+	if not model_cache_path:
+		model_cache_path = model_path
+		clear = True
+	if model_cache_path != model_path:
+		clear = True
+
+	if model_cache and clear is not True:
+		return model_cache
+
+	model_cache = {"groups":[], "objects":[]}
+	if not os.path.isfile(model_path):
+		conf.log("Model path not found")
+		return model_cache
+	if not model_path.lower().endswith('.blend'):
+		conf.log("Model path must be a .blend file")
+		return model_cache
+
+	with bpy.data.libraries.load(model_path) as (data_from, _):
+		if hasattr(data_from, "groups"): # blender 2.7
+			get_attr = "groups"
+		else: # 2.8
+			get_attr = "collections"
+		grp_list = list(getattr(data_from, get_attr))
+		model_cache["groups"] = grp_list
+		for obj in list(data_from.objects):
+			if obj in model_cache["groups"]:
+				# conf.log("Skipping meshwap obj already in cache: "+str(obj))
+				continue
+			# ignore list? e.g. Point.001,
+			model_cache["objects"].append(obj)
+	return model_cache
+
+
+def getModelList(context):
+	"""Only used for UI drawing of enum menus, full list."""
+
+	# may redraw too many times, perhaps have flag
+	if not context.scene.mcprep_props.model_list:
+		updateModelList(context)
+	return [(itm.block, itm.name.title(), "Place {}".format(itm.name))
+			for itm in context.scene.mcprep_props.model_list]
+
+
+def move_assets_to_excluded_layer(context, collections):
```

Comment:
Optional refactor (ie if you don't, I'll do this after); move this function into `util.py`, and have both this module as well as the meshswap module reference this where used, to avoid duplicating the exact code.

Reply:
Will do if I remember

Reply:
Done